ggshield/hmsl/client.py

```diff
@@ -0,0 +1,195 @@
+import base64
+import json
+import re
+from dataclasses import dataclass
+from datetime import datetime, timedelta
+from typing import Any, Dict, Iterable, Iterator, List, Optional, Union, cast, overload
+
+import requests
+
+from .crypto import decrypt, make_hint
+
+
+PREFIX_LENGTH = 5
+HASHES_BATCH_SIZE = 100
+PREFIXES_BATCH_SIZE = 10
+HASH_REGEX = re.compile(r"^[0-9a-f]{64}$")
+PREFIX_REGEX = re.compile(f"^[0-9a-f]{{{PREFIX_LENGTH}}}$")
+
+
+@dataclass
+class Secret:
+    hash: str
+    count: int
+    url: str
+
+
+@dataclass
+class Match:
+    hint: str
+    payload: str
+
+    def decrypt(self, hash: str) -> Secret:
+        key = bytes.fromhex(hash)
+        payload = base64.b64decode(self.payload)
+        decrypted = json.loads(decrypt(payload, key))
+        return Secret(
+            hash=hash,
+            count=decrypted["c"],
+            url=decrypted["l"]["u"],
+        )
+
+
+@dataclass
+class SecretsResponse:
+    secrets: List[Secret]
+
+
+@dataclass
+class PrefixesResponse:
+    matches: List[Match]
+
+
+@dataclass
+class Quota:
+    remaining: int
+    limit: int
+    reset: datetime
+
+
+class HMSLClient:
+    def __init__(
+        self, url: str, jwt: Optional[str] = None, *, prefix_length: int = PREFIX_LENGTH
+    ) -> None:
+        self.url = url.strip("/")
+        self.jwt = jwt
+        self.prefix_length = prefix_length
+        self._quota: Optional[Quota] = None
+
+    @property
+    def quota(self) -> Quota:
+        """Return the remaining credits."""
+        if self._quota is None:
+            # Use the side effect of the call to set the remaining credits
+            self.check_prefixes([])
+        return cast(Quota, self._quota)
+
+    @property
+    def status(self) -> bool:
+        """Return the status of the HMSL server."""
+        try:
+            response = requests.get(f"{self.url}/healthz")
+            response.raise_for_status()
+            return True
+        except requests.exceptions.RequestException:
+            return False
+
+    # Note: these overloads don't perfectly match the implementation,
+    # especially since we don't have Literal annotations in Python 3.7.
+    # But it's good enough for our usage in ggshield.
+    @overload
+    def check(self, hashes: Iterable[str]) -> Iterator[Secret]:
+        ...
+
+    @overload
+    def check(self, hashes: Iterable[str], *, full_hashes: bool) -> Iterator[Secret]:
+        ...
+
+    @overload
+    def check(
+        self, hashes: Iterable[str], *, full_hashes: bool, decrypt: bool
+    ) -> Iterator[Match]:
+        ...
+
+    def check(  # type: ignore
+        self, hashes: Iterable[str], *, full_hashes: bool = False, decrypt: bool = True
+    ):
```

Comment:
I think the code would be nicer if `check()` did not try to do too much. From what I understand, the different use cases are the following:

- The `hmsl check` command calls `check(full_hashes=$arg, decrypt=True)`. It wants Secret instances.

- The `hmsl query` command calls `check(full_hashes=$arg, decrypt=False)`. It only needs dicts, and dumps them as JSON on stdout.


I propose splitting `HMSLClient.check()` in two functions, like this:

```python
class HMSLClient:
    def check(hashes, *, full_hashes: bool) -> Iterable[Secret]
    
    def query(hashes, *, full_hashes: bool) -> Iterable[Dict[str, Any]]
``` 

To avoid implementing the batch iteration twice, we can add a Python 3.7 friendly backport of the upcoming [itertools.batched()](https://docs.python.org/3.12/library/itertools.html#itertools.batched) to core.utils (would be useful in other places too):

```python
def batched(iterable: Iterable[T], batch_size: int) -> Iterable[Tuple[T, ...]]:
    it = iter(iterable)
    while True:
        batch = tuple(islice(it, batch_size))
        if batch:
            yield batch
        else:
            return
```

This makes it possible to implement `HMSLClient.check` like this:

```python
    def check():
        ...
        for batch in batched(hashes, batch_size):
            if full_hashes:
                yield from self.check_hashes(batch).secrets
            else:
                hints = {make_hint(hash): hash for hash in batch}
                for match in response.matches:
                    if match.hint in hints:
                        hash = hints[match.hint]
                        yield match.decrypt(hash)
```                

And `HMSLClient.query` like this:

```python
    def query():
        ...
        for batch in batched(hashes, batch_size):
            if full_hashes:
                yield from map(asdict, self.check_hashes(batch).secrets)
            else:
                yield from map(asdict, response.matches)
```